pycrostates/preprocessing/spatial_filter.py

```diff
@@ -0,0 +1,190 @@
+    * An ajdacency matrix is computed using
+      `mne.channels.find_ch_adjacency`.
+    * If ``exclude_bads`` is set to ``True``,
+      bad channels are removed from the ajdacency matrix.
+    * For each timepoint and each channel,
+      a reduced adjacency matrix is build by removing neighbors
+      with lowest and highest value.
+    * For each timepoint and each channel,
+      a reduced interpolation matrix is built by extracting neighbor
+      weights based on the reduced adjacency matrix.
+    * Reduced interpolation matrices are normalized.
+    * Channel's timepoints are interpolated
+      using their reduced interpolation matrix.
+
+    Parameters
+    ----------
+    inst : Raw | Epochs | ChData
+        Instance to spatial filter.
+    ch_type : str
+        The channel type on which to apply spatial filter.
+        Currently only supports ``'eeg'``.
+        Default to 'eeg'.
+    exclude_bads : bool
+        If set to ``True`` bad channels will be removed
+        from the adjacency matrix and therefore not be used
+        to interpolate neighbors. In addition, bad channels
+        will not be filtered.
+        If set to ``False``, proceed as if all channels were good.
+        Default to ``True``.
+    origin : array-like, shape (3,) | str
+        Origin of the sphere in the head coordinate frame and in meters.
+        Can be ``'auto'`` (default), which means a head-digitization-based
+        origin fit.
+    %(n_jobs)s
+    %(verbose)s
+
+    Returns
+    -------
+    inst : Raw | Epochs| ChData
+        The modified instance.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    # Checks
+    from ..io import ChData
+
+    _check_type(inst, (BaseRaw, BaseEpochs, ChData), item_name="inst")
+    _check_value(ch_type, ("eeg",), item_name="ch_type")
+    n_jobs = _check_n_jobs(n_jobs)
+    # check preload for Raw
+    _check_preload(inst, "Apply spatial filter")
+    # check montage
+    if inst.get_montage() is None:
+        raise ValueError(
+            "No montage was set on your data, but spatial filter"
+            "can only work if digitization points for the EEG "
+            "channels are available. Consider calling inst.set_montage() "
+            "to apply a montage."
+        )
+    # remove bad channels
+    # Extract adjacency matrix
+    adjacency_matrix, ch_names = mne.channels.find_ch_adjacency(
+        inst.info, ch_type
+    )
+    adjacency_matrix = adjacency_matrix.todense()
+    adjacency_matrix = np.array(adjacency_matrix)
+    if exclude_bads:
+        for c, chan in enumerate(ch_names):
+            if chan in inst.info["bads"]:
+                adjacency_matrix[c, :] = 0  # do not change bads
+                adjacency_matrix[:, c] = 0  # don't use bads to interpolate
+    # retrieve picks based on adjacency matrix
+    picks = dict(_picks_by_type(inst.info, exclude=[]))[ch_type]
+    info = mne.pick_info(inst.info, picks)
+    assert ch_names == info.ch_names
```

Comment:
This is a sanity-check and not something that is ever expected to fail due to the provided object from the user? In which case it would have to be a better error with an informative error message.

Reply:
Yes this is a sanity check.